# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: js/src/frontend/TryEmitter.cpp
# Commit: 5f6b9e2fc3bb
# Full Hash: 5f6b9e2fc3bb3c5fd74f9871399269a2665a5270
# Author: Iain Ireland <iireland@mozilla.com>
# Date: 2022-03-09 09:44:44
# Regressor Bug: 885514
# File Overlap Count: 1
# Description:
#   Bug 885514: Refactor try/finally to use goto r=jandem
#   
#   JSOp::Gosub is just a JSOp::Goto that lies about its effect on the stack. We can start simplifying try/finally by using a JSOp::Goto instead.
#   
#   Currently, to make the stack depth work out when returning from a finally, Gosub pretends to pop two values (the arguments to retsub), and Finally pretends to push them. This means that bytecode analysis can treat Gosub as if it falls through, and the stack depth of the jumptarget following the gosub will be correct. This patch uses a different approach: bytecode analysis marks the return address as reachable when we push it using JSOp::ResumeIndex. This means that the bytecode no longer has to lie about stack depth.
# ==============================================================================

diff -r b199b21ed5a3 -r 5f6b9e2fc3bb js/src/frontend/TryEmitter.cpp
--- a/js/src/frontend/TryEmitter.cpp	Tue Mar 08 21:00:30 2022 +0000
+++ b/js/src/frontend/TryEmitter.cpp	Tue Mar 08 21:10:21 2022 +0000
@@ -61,9 +61,9 @@
   MOZ_ASSERT(state_ == State::Try);
   MOZ_ASSERT(depth_ == bce_->bytecodeSection().stackDepth());
 
-  // Gosub to finally, if present.
+  // Jump to finally, if present.
   if (hasFinally() && controlInfo_) {
-    if (!bce_->emitGoSub(&controlInfo_->gosubs)) {
+    if (!bce_->emitJumpToFinally(&controlInfo_->finallyJumps_)) {
       return false;
     }
   }
@@ -118,9 +118,9 @@
     return true;
   }
 
-  // gosub <finally>, if required.
+  // Jump to <finally>, if required.
   if (hasFinally()) {
-    if (!bce_->emitGoSub(&controlInfo_->gosubs)) {
+    if (!bce_->emitJumpToFinally(&controlInfo_->finallyJumps_)) {
       return false;
     }
     MOZ_ASSERT(bce_->bytecodeSection().stackDepth() == depth_);
@@ -164,14 +164,19 @@
 
   MOZ_ASSERT(bce_->bytecodeSection().stackDepth() == depth_);
 
+  // Upon entry to the finally, there are two additional values on the stack:
+  // a boolean value to indicate whether we're throwing an exception, and
+  // either that exception (if we're throwing) or a resume index to which we
+  // will return (if we're not throwing).
+  bce_->bytecodeSection().setStackDepth(depth_ + 2);
+
   if (!bce_->emitJumpTarget(&finallyStart_)) {
     return false;
   }
 
   if (controlInfo_) {
-    // Fix up the gosubs that might have been emitted before non-local
-    // jumps to the finally code.
-    bce_->patchJumpsToTarget(controlInfo_->gosubs, finallyStart_);
+    // Fix up the jumps to the finally code.
+    bce_->patchJumpsToTarget(controlInfo_->finallyJumps_, finallyStart_);
 
     // Indicate that we're emitting a subroutine body.
     controlInfo_->setEmittingSubroutine();